include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,501 @@
+    {
+        none,
+        valid,
+        invalid
+    };
+
+    struct polygon_projection
+    {
+        polygon_2d projected_polygon;
+        vec3 base;
+        vec3 v1;
+        vec3 v2;
+        vec3 normal;
+    };
+
+    template <int I, int J, typename Strategy>
+    static bool affinely_independent(point_3d const& p1, point_3d const& p2, point_3d const& p3,
+                                     Strategy const& strategy)
+    {
+        point_2d const p1_2{geometry::get<I>(p1), geometry::get<J>(p1)};
+        point_2d const p2_2{geometry::get<I>(p2), geometry::get<J>(p2)};
+        point_2d const p3_2{geometry::get<I>(p3), geometry::get<J>(p3)};
+        return strategy.side().apply(p1_2, p2_2, p3_2) != 0;
+    }
+
+    // Try to find 3 non-collinear points
+    template <typename Strategy>
+    static bool find_non_collinear_points(polygon_3d const& polygon, vec3& base, vec3& v1, vec3& v2,
+                                          Strategy const& strategy)
+    {
+        for (auto pit1 = points_begin(polygon); pit1 != points_end(polygon); ++pit1)
+        {
+            for (auto pit2 = pit1; pit2 != points_end(polygon); ++pit2)
+            {
+                for (auto pit3 = pit2; pit3 != points_end(polygon); ++pit3)
+                {
+                    if (affinely_independent<0, 1>(*pit1, *pit2, *pit3, strategy)
+                        || affinely_independent<0, 2>(*pit1, *pit2, *pit3, strategy)
+                        || affinely_independent<1, 2>(*pit1, *pit2, *pit3, strategy))
+                    {
+                        set_vec_from_point::apply(*pit1, base);
+                        set_vec_from_point::apply(*pit2, v1);
+                        set_vec_from_point::apply(*pit3, v2);
+                        return true;
+                    }
+                }
+            }
+        }
+        return false; // All points are collinear
+    }
+
+    static point_2d project_point_to_plane_2d(vec3 const& v1, vec3 const& v2, vec3 const& p)
+    {
+        return point_2d{qvm::dot(v1, p), qvm::dot(v2, p)};
+    }
+
+    template <typename VisitPolicy>
+    static bool has_correct_intersection_orientation(polygon_2d const& polygon,
+                                                     point_2d const& point1,
+                                                     point_2d const& point2,
+                                                     VisitPolicy& visitor)
+    {
+        auto ring = boost::geometry::exterior_ring(polygon);
+        if (boost::geometry::closure<ring_2d>::value == boost::geometry::closed)
+        {
+            ring.pop_back(); // Remove the closing point to make it open
+        }
+
+        for (auto it = boost::begin(ring); it != boost::end(ring); ++it)
+        {
+            if (boost::geometry::equals(*it, point1))
+            {
+                if (boost::geometry::equals(point1, point2))
+                {
+                    visitor.template apply<failure_duplicate_points>();
```

Comment:
Can't this check be made right at the beginning, before even iterating?
What are point1 and point2? Is it a segment?
Can we then rename them to `segment_point1` or just `s1` with a comment?